File: html/en/7gui/circle_drawer.c
#include <stdlib.h>
#include <math.h>
#include <iup.h>
#include <iupdraw.h>


/*********************************** Circle List ****************************************/

#define DEF_RADIUS 30
#define MAX_RADIUS 100
#define MAX_CIRCLES 80

typedef struct _Circle
{
  int x;
  int y;
  int r;
  int id;
} Circle;

static Circle circlesList[MAX_CIRCLES];
static int circlesCount = 0;

static int circles_add(int x, int y)
{
  static int lastId = 0;

  if (circlesCount == MAX_CIRCLES)
    return -1;

  circlesList[circlesCount].x = x;
  circlesList[circlesCount].y = y;
  circlesList[circlesCount].r = DEF_RADIUS;
  circlesList[circlesCount].id = lastId;  /* each circle has an unique id */

  circlesCount++;
  lastId++;
  return lastId - 1;
}

static int circles_insert(Circle circle)
{
  if (circlesCount == MAX_CIRCLES)
    return -1;

  circlesList[circlesCount] = circle;
  circlesCount++;
  return 1;
}

static void circles_remove(Circle circle)
{
  int i, j;
  for (i = 0; i < circlesCount; i++)
  {
    if (circlesList[i].id == circle.id)
      break;
  }

  for (j = i; j < circlesCount - 1; j++)
    circlesList[j] = circlesList[j + 1];

  circlesCount--;
}

static void circles_update(Circle circle)
{
  int i;
  for (i = 0; i < circlesCount; i++)
  {
    if (circlesList[i].id == circle.id)
      break;
  }

  circlesList[i].r = circle.r;
}

static int circles_pick(int x, int y)
{
  int i;
  int min_r = MAX_RADIUS;
  int id = -1;

  for (i = 0; i < circlesCount; i++)
  {
    int dx = circlesList[i].x - x;
    int dy = circlesList[i].y - y;
    int r = dx*dx + dy*dy;

    r = (int)sqrt((double)r);

    if (r < circlesList[i].r)
    {
      if (id < 0)
      {
        id = circlesList[i].id;
        min_r = r;
      }
      else if (r < min_r)
        id = circlesList[i].id;
    }
  }

  return id;
}

static Circle circles_find(int id)
{
  int i;
  for (i = 0; i < circlesCount; i++)
  {
    if (circlesList[i].id == id)
      break;
  }
  return circlesList[i];
}


/************************************* Operation List ************************************************/

#define MAX_OPERATIONS 200

enum { OP_CREATE, OP_REMOVE, OP_UPDATE };

typedef struct _Operation
{
  int type;
  Circle circle;
} Operation;

static Operation operationsList[MAX_OPERATIONS];
static int operationsCount = 0;
static int currentOperation = -1;

static void op_add(int type, Circle circle)
{
  if (operationsCount == MAX_OPERATIONS)
    return;

  if (currentOperation < operationsCount - 1)
    operationsCount = currentOperation + 1;

  operationsList[operationsCount].type = type;
  operationsList[operationsCount].circle = circle;
  currentOperation = operationsCount;

  operationsCount++;
}

static void op_undo(void)
{
  Operation op;

  if (currentOperation < 0)
    return;

  op = operationsList[currentOperation];
  currentOperation--;

  switch (op.type)
  {
    case OP_CREATE:
      circles_remove(op.circle);
      break;
    case OP_REMOVE:
      circles_insert(op.circle);
      break;
    case OP_UPDATE:
      op = operationsList[currentOperation];
      currentOperation--; /* return 2 to get the old value */

      circles_update(op.circle);
      break;
  }
}

static void op_redo(void)
{
  Operation op;

  if (currentOperation + 1 == operationsCount)
    return;

  currentOperation++;
  op = operationsList[currentOperation];

  switch (op.type)
  {
    case OP_CREATE:
      circles_insert(op.circle);
      break;
    case OP_REMOVE:
      circles_remove(op.circle);
      break;
    case OP_UPDATE:
      currentOperation++; /* increment 2 to get the new value */
      op = operationsList[currentOperation];

      circles_update(op.circle);
      break;
  }
}

static int op_has_undo(void)
{
  if (currentOperation < 0)
    return 0;

  return 1;
}

static int op_has_redo(void)
{
  if (currentOperation + 1 == operationsCount)
    return 0;

  return 1;
}


/*********************************************** Interface ****************************************/

void update_buttons(Ihandle* ih)
{
  Ihandle *undoButton = IupGetDialogChild(ih, "UNDO");
  Ihandle *redoButton = IupGetDialogChild(ih, "REDO");

  if (op_has_undo())
    IupSetAttribute(undoButton, "ACTIVE", "YES");
  else
    IupSetAttribute(undoButton, "ACTIVE", "NO");

  if (op_has_redo())
    IupSetAttribute(redoButton, "ACTIVE", "YES");
  else
    IupSetAttribute(redoButton, "ACTIVE", "NO");
}

int bt_undo_cb(Ihandle *self)
{
  Ihandle *canvas = IupGetDialogChild(self, "CANVAS");

  op_undo();

  update_buttons(self);

  IupUpdate(canvas);

  return IUP_DEFAULT;
}

int bt_redo_cb(Ihandle *self)
{
  Ihandle *canvas = IupGetDialogChild(self, "CANVAS");

  op_redo();

  update_buttons(self);

  IupUpdate(canvas);

  return IUP_DEFAULT;
}

static int canvas_button_cb(Ihandle *ih, int but, int pressed, int x, int y, char* status)
{
  Ihandle *dlg = IupGetDialog(ih);

  if (but == IUP_BUTTON1 && pressed)
  {
    int id = circles_pick(x, y);
    if (id >= 0)
      IupSetInt(ih, "HIGHLIGHTEDCIRCLE", id);
    else
    {
      id = circles_add(x, y);
      op_add(OP_CREATE, circles_find(id));

      update_buttons(ih);
    }

    IupUpdate(ih);
  }
  else if (but == IUP_BUTTON3 && pressed)
  {
    int highlightedCircle = IupGetInt(ih, "HIGHLIGHTEDCIRCLE");
    int id = circles_pick(x, y);
    if (id == highlightedCircle)
    {
      Circle circle = circles_find(id);
      int new_r, old_r = circle.r;
      Ihandle *configDialog = (Ihandle *)IupGetAttribute(dlg, "CONFIGDIALOG");
      Ihandle *val = IupGetDialogChild(configDialog, "VAL");
      Ihandle *lbl = IupGetDialogChild(configDialog, "LBL");

      IupSetStrf(lbl, "TITLE", "Adjust the diameter of the circle at (%d, %d)", circle.x, circle.y);
      IupSetInt(val, "VALUE", old_r);
      IupSetInt(configDialog, "CIRCLEID", id);

      IupPopup(configDialog, IUP_CENTERPARENT, IUP_CENTERPARENT);

      new_r = IupGetInt(val, "VALUE");
      if (new_r != old_r)
      {
        op_add(OP_UPDATE, circle); /* add both old and new */
        circle.r = new_r;
        op_add(OP_UPDATE, circle);

        update_buttons(ih);
      }
    }
  }

  (void)status;
  return IUP_DEFAULT;
}

static int canvas_action(Ihandle *ih)
{
  int i, w, h, highlightedCircle;

  IupDrawBegin(ih);

  IupDrawGetSize(ih, &w, &h);

  IupSetAttribute(ih, "DRAWCOLOR", "255 255 255");
  IupSetAttribute(ih, "DRAWSTYLE", "FILL");
  IupDrawRectangle(ih, 0, 0, w - 1, h - 1);

  IupSetAttribute(ih, "DRAWCOLOR", "0 0 0");
  IupSetAttribute(ih, "DRAWSTYLE", "STROKE");

  highlightedCircle = IupGetInt(ih, "HIGHLIGHTEDCIRCLE");

  for (i = 0; i < circlesCount; i++)
  {
    Circle circle = circlesList[i];

    if (highlightedCircle == circle.id)
    {
      IupSetAttribute(ih, "DRAWCOLOR", "128 128 128");
      IupSetAttribute(ih, "DRAWSTYLE", "FILL");
    }
    else
    {
      IupSetAttribute(ih, "DRAWCOLOR", "0 0 0");
      IupSetAttribute(ih, "DRAWSTYLE", "STROKE");
    }

    IupDrawArc(ih, circle.x - circle.r, circle.y - circle.r,
                   circle.x + circle.r, circle.y + circle.r, 0., 360.);
  }

  IupDrawEnd(ih);
  return IUP_DEFAULT;
}

int val_valuechanged_cb(Ihandle *self)
{
  Ihandle* dlg = IupGetDialog(self);
  Ihandle* canvas = (Ihandle *)IupGetAttribute(self, "CANVAS");
  int circleId = IupGetInt(dlg, "CIRCLEID");
  int r = IupGetInt(self, "VALUE");
  Circle circle = circles_find(circleId);

  circle.r = r;
  circles_update(circle);

  IupUpdate(canvas);

  return IUP_DEFAULT;
}

Ihandle *createCircleConfigDialog(Ihandle* parent_dlg)
{
  Ihandle *dlg, *val, *lbl, *box;

  lbl = IupLabel(NULL);
  IupSetAttribute(lbl, "EXPAND", "HORIZONTAL");
  IupSetAttribute(lbl, "TITLE", "Adjust the diameter of the circle at (100, 100)");
  IupSetAttribute(lbl, "NAME", "LBL");

  val = IupVal("HORIZONTAL");
  IupSetAttribute(val, "EXPAND", "HORIZONTAL");
  IupSetAttribute(val, "MIN", "0");
  IupSetInt(val, "MAX", MAX_RADIUS);
  IupSetAttribute(val, "NAME", "VAL");

  box = IupVbox(IupFill(), lbl, val, IupFill(), NULL);
  IupSetAttribute(box, "ALIGMENT", "ACENTER");

  IupSetAttribute(box, "NMARGIN", "10x10");
  dlg = IupDialog(box);
  IupSetAttribute(dlg, "TITLE", "Circle Config");
  IupSetAttributeHandle(dlg, "PARENTDIALOG", parent_dlg);

  IupSetCallback(val, "VALUECHANGED_CB", (Icallback)val_valuechanged_cb);

  return dlg;
}

int main(int argc, char **argv)
{
  Ihandle *dlg, *hbox, *vbox, *configDialog;
  Ihandle *undoButton, *redoButton, *canvas;

  IupOpen(&argc, &argv);

  undoButton = IupButton("Undo", NULL);
  IupSetAttribute(undoButton, "NAME", "UNDO");
  IupSetAttribute(undoButton, "SIZE", "60");
  redoButton = IupButton("Redo", NULL);
  IupSetAttribute(redoButton, "NAME", "REDO");
  IupSetAttribute(redoButton, "SIZE", "60");

  IupSetAttribute(redoButton, "ACTIVE", "NO");
  IupSetAttribute(undoButton, "ACTIVE", "NO");

  canvas = IupCanvas(NULL);
  IupSetAttribute(canvas, "NAME", "CANVAS");
  IupSetAttribute(canvas, "EXPAND", "YES");
  IupSetInt(canvas, "HIGHLIGHTEDCIRCLE", -1);

  hbox = IupHbox(IupFill(), undoButton, redoButton, IupFill(), NULL);

  vbox = IupVbox(hbox, canvas, NULL);

  IupSetAttribute(vbox, "NMARGIN", "10x10");
  IupSetAttribute(vbox, "GAP", "10");
  dlg = IupDialog(vbox);
  IupSetAttribute(dlg, "TITLE", "Circle Drawer");
  IupSetAttribute(dlg, "SIZE", "300x150");

  /* Registers callbacks */
  IupSetCallback(undoButton, "ACTION", (Icallback)bt_undo_cb);
  IupSetCallback(redoButton, "ACTION", (Icallback)bt_redo_cb);
  IupSetCallback(canvas, "ACTION", (Icallback)canvas_action);
  IupSetCallback(canvas, "BUTTON_CB", (Icallback)canvas_button_cb);

  configDialog = createCircleConfigDialog(dlg);

  IupSetAttribute(configDialog, "CANVAS", (char *)canvas);

  IupSetAttribute(dlg, "CONFIGDIALOG", (char *)configDialog);

  IupShowXY(dlg, IUP_CENTER, IUP_CENTER);

  IupMainLoop();

  IupClose();
  return EXIT_SUCCESS;
}

File: html/en/7gui/circle_drawer.lua
require("iuplua")


--*********************************** Circle List ****************************************/

DEF_RADIUS = 30
MAX_RADIUS = 100
circlesList = {}
circlesCount = 0
lastId = 1

function circles_add(x, y)
	circlesCount = circlesCount + 1
	circlesList[circlesCount] = {}
	circlesList[circlesCount].x = x
	circlesList[circlesCount].y = y
	circlesList[circlesCount].r = DEF_RADIUS
	circlesList[circlesCount].id = lastId  -- each circle has an unique id
	lastId = lastId + 1
	return lastId - 1
end

function circles_insert(id, x, y, r)
	circlesCount = circlesCount + 1
	circlesList[circlesCount] = {}
	circlesList[circlesCount].x = x
	circlesList[circlesCount].y = y
	circlesList[circlesCount].r = r
	circlesList[circlesCount].id = id
	return 1
end

function circles_remove(id)
	local index
	for i = 1, circlesCount do
		if circlesList[i].id == id then
			index = i
			break
		end
	end
	for j = index, circlesCount - 1 do
		circlesList[j] = circlesList[j + 1]
	end
	circlesCount = circlesCount - 1
end

function circles_update(id, r)
	local index
	for i = 1, circlesCount do
		if circlesList[i].id == id then
			index = i
			break
		end
	end
	circlesList[index].r = r
end

function circles_pick(x, y)
	local min_r = MAX_RADIUS
	local id = -1
	for i = 1, circlesCount do
		local dx = circlesList[i].x - x
		local dy = circlesList[i].y - y
		local r = dx*dx + dy*dy
		r = r^0.5
		if r < circlesList[i].r then
			if id < 0 then
				id = circlesList[i].id
				min_r = r
			elseif r < min_r then
				id = circlesList[i].id
			end
		end
	end
	return id
end

function circles_find(id)
	local index
	for i = 1, circlesCount do
		if circlesList[i].id == id then
			index = i
			break
		end
	end
	return circlesList[index]
end

--************************************* Operation List ************************************************/

MAX_OPERATIONS = 200

OP_CREATE = 1
OP_REMOVE = 2
OP_UPDATE = 3

operationsList = {}
operationsCount = 0
currentOperation = 0

function op_add(op_type, id, x, y, r)
	if currentOperation < operationsCount then
		operationsCount = currentOperation
	end
	operationsCount = operationsCount + 1
	operationsList[operationsCount] = {}
	operationsList[operationsCount].type = op_type
	operationsList[operationsCount].x = x
	operationsList[operationsCount].y = y
	operationsList[operationsCount].r = r
	operationsList[operationsCount].id = id
	currentOperation = operationsCount
end

function op_undo(void)
	if currentOperation < 1 then
		return
	end

	local op = operationsList[currentOperation]
	currentOperation = currentOperation - 1

	if op.type == OP_CREATE then
		circles_remove(op.id)
	elseif op.type == OP_REMOVE then
		circles_insert(op.id, op.x, op.y, op.r)
	elseif op.type == OP_UPDATE then
		op = operationsList[currentOperation]
		currentOperation = currentOperation - 1 -- return 2 to get the old value 

		circles_update(op.id, op.r)
	end
end

function op_redo(void)
	if currentOperation == operationsCount then
		return
	end

	currentOperation = currentOperation + 1
	op = operationsList[currentOperation]

	if op.type == OP_CREATE then
		circles_insert(op.id, op.x, op.y, op.r)
	elseif op.type == OP_REMOVE then
		circles_remove(op.id)
	elseif op.type == OP_UPDATE then
		currentOperation = currentOperation + 1 -- increment 2 to get the new value
		op = operationsList[currentOperation]

		circles_update(op.id, op.r)
	end
end

function op_has_undo()
	if currentOperation < 1 then
		return nil
	end
	return 1
end

function op_has_redo()
	if currentOperation == operationsCount then
		return nil
	end
	return 1
end


--*********************************************** Interface ****************************************/

function update_buttons()
	if op_has_undo() then
		undoButton.active = "YES"
	  else
		undoButton.active = "NO"
	end

	if op_has_redo() then
		redoButton.active = "YES"
	  else
		redoButton.active = "NO"
	end
end

undoButton = iup.button{title = "Undo", size = "60", active = "NO"}
redoButton = iup.button{title = "Redo", size = "60", active = "NO"}

canvas = iup.canvas{expand = "YES", highlightedCircle = -1}

hbox = iup.hbox{iup.fill{}, undoButton, redoButton, iup.fill{}}

vbox = iup.vbox{hbox, canvas;  nmargin = "10x10", gap = "10"}

dlg = iup.dialog{vbox, title = "Circle Drawer", size = "300x150"}

val_lbl = iup.label{"Adjust the diameter of the circle at (100, 100)"; expand = "HORIZONTAL"}

val = iup.val{"HORIZONTAL"; expand = "HORIZONTAL", min = "0", max = MAX_RADIUS}

config_box = iup.vbox{iup.fill{}, val_lbl, val, iup.fill{}, alignment = "ACENTER", nmargin = "10x10"}

config_dial = iup.dialog{config_box; parentdialog = dlg, title = "Circle Config"}

function undoButton:action()
	op_undo()
	update_buttons()
	iup.Update(canvas)
end

function redoButton:action()
	op_redo()
	update_buttons()
	iup.Update(canvas)
end

function canvas:action(posx, posy)
	iup.DrawBegin(self)
	local w, h = iup.DrawGetSize(self)
	self.drawcolor = "255 255 255"
	self.drawstyle = "FILL"
	iup.DrawRectangle(self, 0, 0, w - 1, h - 1)
	self.drawcolor = "0 0 0"
	self.drawstyle = "STROKE"
	local highlightedCircle = tonumber(self.highlightedcircle)
	for i = 1, circlesCount do
    local circle = circlesList[i]
		if highlightedCircle == circle.id then
			self.drawcolor = "128 128 128"
			self.drawstyle = "FILL"
		else
			self.drawcolor = "0 0 0"
			self.drawstyle = "STROKE"
		end
		iup.DrawArc(self, circle.x - circle.r, circle.y - circle.r, 
                      circle.x + circle.r, circle.y + circle.r, 0., 360.)
	end
	iup.DrawEnd(self)
end

function canvas:button_cb(button, pressed, x, y, status)
	if button == iup.BUTTON1 and pressed == 1 then
		local id = circles_pick(x, y)
		if id > 0 then
			self.highlightedcircle = id
		else
			id = circles_add(x, y)
			local circle = circles_find(id)
			op_add(OP_CREATE, circle.id, circle.x, circle.y, circle.r)
			update_buttons()
		end
		iup.Update(self)
	elseif button == iup.BUTTON3 and pressed == 1 then
		local highlightedCircle = tonumber(self.highlightedcircle)
		local id = circles_pick(x, y)
		if id == highlightedCircle then
			local circle = circles_find(id)
			local old_r = circle.r
			local new_r = old_r
			val_lbl.title = string.format("Adjust the diameter of the circle at (%d, %d)", circle.x, circle.y)
			val.value = old_r
			config_dial.circle_id = id
			iup.Popup(config_dial, iup.CENTERPARENT, iup.CENTERPARENT)
			new_r = tonumber(val.value)
			if new_r ~= old_r then
				op_add(OP_UPDATE, circle.id, circle.x, circle.y, old_r) -- add both old and new
				op_add(OP_UPDATE, circle.id, circle.x, circle.y, new_r)
				update_buttons()
			end
		end
	end
end

function val:valuechanged_cb()
	local circle_id = tonumber(config_dial.circle_id)
	local r = tonumber(self.value)
	circles_update(circle_id, r)
	iup.Update(canvas)
end

dlg:showxy( iup.CENTER, iup.CENTER )

if (iup.MainLoopLevel()==0) then
	iup.MainLoop()
end
